types/coin.go

```diff
@@ -110,118 +127,169 @@ func (coins Coins) IsValid() bool {
 				return false
 			}
+
 			// we compare each coin against the last denom
 			lowDenom = coin.Denom
 		}
+
 		return true
 	}
 }
 
-// Plus combines two sets of coins
-// CONTRACT: Plus will never return Coins where one Coin has a 0 amount.
+// Plus adds two sets of coins.
+//
+// e.g.
+// {2A} + {A, 2B} = {3A, 2B}
+// {2A} + {0B} = {2A}
+//
+// NOTE: Plus uses the invariant that coins are sorted by denominations.
+//
+// CONTRACT: Plus will never return Coins where one Coin has a negative amount.
 func (coins Coins) Plus(coinsB Coins) Coins {
+	return coins.safePlus(coinsB)
+}
+
+// safePlus will perform addition of two coins sets. If both coin sets are
+// empty, then an empty set is returned. If only a single set is empty, the
+// other set is returned. Otherwise, the coins are compared in order of their
+// denomination and addition only occurs when the denominations match, otherwise
+// the coin is simply added to the sum assuming it's not zero.
+func (coins Coins) safePlus(coinsB Coins) Coins {
 	sum := ([]Coin)(nil)
 	indexA, indexB := 0, 0
 	lenA, lenB := len(coins), len(coinsB)
+
 	for {
 		if indexA == lenA {
 			if indexB == lenB {
+				// return nil coins if both sets are empty
 				return sum
 			}
-			return append(sum, coinsB[indexB:]...)
+
+			// return set B (excluding zero coins) if set A is empty
+			return append(sum, removeZeroCoins(coinsB[indexB:])...)
 		} else if indexB == lenB {
-			return append(sum, coins[indexA:]...)
+			// return set A (excluding zero coins) if set B is empty
+			return append(sum, removeZeroCoins(coins[indexA:])...)
 		}
+
 		coinA, coinB := coins[indexA], coinsB[indexB]
+
 		switch strings.Compare(coinA.Denom, coinB.Denom) {
-		case -1:
+		case -1: // coin A denom < coin B denom
 			if coinA.IsZero() {
-				// ignore 0 sum coin type
+				// ignore 0 sum coin
 			} else {
 				sum = append(sum, coinA)
 			}
+
 			indexA++
-		case 0:
-			if coinA.Amount.Add(coinB.Amount).IsZero() {
+
+		case 0: // coin A denom == coin B denom
+			res := coinA.Plus(coinB)
+			if res.IsZero() {
 				// ignore 0 sum coin type
 			} else {
-				sum = append(sum, coinA.Plus(coinB))
+				sum = append(sum, res)
 			}
+
 			indexA++
 			indexB++
-		case 1:
+
+		case 1: // coin A denom > coin B denom
 			if coinB.IsZero() {
-				// ignore 0 sum coin type
+				// ignore 0 sum coin
 			} else {
 				sum = append(sum, coinB)
 			}
+
 			indexB++
 		}
 	}
 }
 
-// Negative returns a set of coins with all amount negative
-func (coins Coins) Negative() Coins {
-	res := make([]Coin, 0, len(coins))
-	for _, coin := range coins {
-		res = append(res, Coin{
-			Denom:  coin.Denom,
-			Amount: coin.Amount.Neg(),
-		})
+// Minus subtracts a set of coins from another.
+//
+// e.g.
+// {2A, 3B} - {A} = {A, 3B}
+// {2A} - {0B} = {2A}
+// {A, B} - {A} = {B}
+//
+// CONTRACT: Minus will never return Coins where one Coin has a negative amount.
+func (coins Coins) Minus(coinsB Coins) Coins {
+	res := coins.safePlus(coinsB.negative())
```

Comment:
This is still not a great way to subtract coins (it wasn't before either), unnecessary allocation of a new `Coins` array

Doesn't necessarily need to be fixed in this PR

Reply:
Will now modify `coinsB` in place during `Minus`.

Reply:
Suggestions on how to improve this w/o modifying `coinsB`?

Reply:
Iterate through the sorted arrays and subtract amounts (still allocating a new array, but this version allocates two: `coinsB.negative()` and then the result).

Reply:
@cwgoes upon further attempt, I can figure out a way to do this, but would rather have it in a separate PR..maybe as part of the switch to uints.